scylla/src/observability/metrics.rs

```diff
@@ -1,23 +1,204 @@
+        Self {
+            one_minute_rate: ExponentiallyWeightedMovingAverage::new(
+                1.0 - (-(INTERVAL as f64) / 60.0 / 1.0).exp(),
+            ),
+            five_minute_rate: ExponentiallyWeightedMovingAverage::new(
+                1.0 - (-(INTERVAL as f64) / 60.0 / 5.0).exp(),
+            ),
+            fifteen_minute_rate: ExponentiallyWeightedMovingAverage::new(
+                1.0 - (-(INTERVAL as f64) / 60.0 / 15.0).exp(),
+            ),
+            count: AtomicU64::new(0),
+            start_time: now,
+            last_tick: AtomicU64::new(now.elapsed().as_nanos() as u64),
+        }
+    }
+
+    fn mark(&self) {
+        self.tick_if_necessary();
+        self.count.fetch_add(1, ORDER_TYPE);
+        self.one_minute_rate.update();
+        self.five_minute_rate.update();
+        self.fifteen_minute_rate.update();
+    }
+
+    fn one_minute_rate(&self) -> f64 {
+        self.one_minute_rate.rate()
+    }
+
+    fn five_minute_rate(&self) -> f64 {
+        self.five_minute_rate.rate()
+    }
+
+    fn fifteen_minute_rate(&self) -> f64 {
+        self.fifteen_minute_rate.rate()
+    }
+
+    fn mean_rate(&self) -> f64 {
+        let count = self.count();
+        if count == 0 {
+            0.0
+        } else {
+            let elapsed = self.start_time.elapsed().as_secs_f64();
+            count as f64 / elapsed
+        }
+    }
+
+    fn count(&self) -> u64 {
+        self.count.load(ORDER_TYPE)
+    }
+
+    fn tick_if_necessary(&self) {
+        // Multiple threads could read the same `old_tick`...
+        let old_tick = self.last_tick.load(ORDER_TYPE);
+        let new_tick = self.start_time.elapsed().as_nanos() as u64;
+        let elapsed = new_tick - old_tick;
+
+        // _"Problematic"_ `if` - see a comment below.
+        if elapsed > INTERVAL * 1_000_000_000 {
+            let new_interval_start_tick = new_tick - elapsed % (INTERVAL * 1_000_000_000);
+            // But then only one will succeed in the following COMPARE EXCHANGE operation.
+            if self
+                .last_tick
+                .compare_exchange(old_tick, new_interval_start_tick, ORDER_TYPE, ORDER_TYPE)
+                .is_ok()
+            {
+                let required_ticks = elapsed / (INTERVAL * 1_000_000_000);
+                // So only one thread will do the following ticks.
+                // My only concern is that this loop might take so long that another thread
+                // enters the _"problematic"_ `if` and then we have a logical race there.
+                // BUT this is extremely unlikely, because then the loop would have to take
+                // 5 seconds! (INTERVAL * 1e9). So I think we are safe from those races.
+                for _ in 0..required_ticks {
+                    self.one_minute_rate.tick();
+                    self.five_minute_rate.tick();
+                    self.fifteen_minute_rate.tick();
+                }
+            }
+        }
+    }
+}
```

Comment:
:thought_balloon: I believe the body of this function could be put into `unsafe` block in order to warn readers and further implementers, and to allow calling `tick()` after making it an `unsafe` function.